sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -141,8 +142,73 @@ trait ScalaReflection {
     ObjectType(cls)
   }
 
+  def constructorFor(schema: StructType): Expression = {
```

Comment:
I'd consider just making an object called `RowEncoder` and put all of the code there (including the helper functions in `EncodeUtil`).  Reasoning being: none of this uses mirrors, types or any of the other scala reflection stuff and the code in EncodeUtil is pretty specific to Rows so doesn't need to be in its own object.